[WIP] Feature/resize images

yachtcaptain23 opened the pull request:
Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave-intl/publishers/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Tagged reviewers.

Test Plan:


Reviewer Checklist:

Tests
- [ ] Adequate test coverage exists to prevent regressions

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))

yachtcaptain23 committed:
Include mini_magick

yachtcaptain23 committed:
Convert to single image size
Use uniform file size

yachtcaptain23 committed:
Initial commit for site banner resizing logic

yachtcaptain23 committed:
Fix background saving by limiting file size

yachtcaptain23 committed:
Add tests for padding logo. Increased number of resizing loops

yachtcaptain23 committed:
Test to make sure hashing is always consistent

nvonpentz commented on app/helpers/image_conversion_helper.rb:
```diff
@@ -0,0 +1,84 @@
+=end
+  def add_padding_to_image(source_image_path:, attachment_type:)
+    # Add initial conversion to get file size
+    iteration = 0
```
can we initialize `iteration` just above the while loop?

nvonpentz commented on app/controllers/publishers/site_banners_controller.rb:
```diff
@@ -39,32 +40,53 @@ def update_background_image
-    file = Tempfile.new([filename, extension])
-    File.open(file.path, 'wb') do |f|
+    temp_file = Tempfile.new([filename, extension])
+    File.open(temp_file.path, 'wb') do |f|
```
should we explicitly close this tempfile?

yachtcaptain23 replied:
Per https://ruby-doc.org/core-2.5.1/File.html#M000069 and https://stackoverflow.com/questions/4795447/rubys-file-open-and-the-need-for-f-close

`If the optional code block is given, it will be passed the opened file as an argument and the File object will automatically be closed when the block terminates.`

nvonpentz commented on app/models/site_banner.rb:
```diff
@@ -7,6 +7,14 @@ class SiteBanner < ApplicationRecord
+
+  BACKGROUND = "background".freeze
+  BACKGROUND_DIMENSIONS = [900,176]
+  BACKGROUND_UNIVERSAL_FILE_SIZE = 60_000
```
can you add some comments here, above, and for `MAX_IMAGE_SIZE` that specify the unit

nvonpentz reviewed (commented):
Left some minor comments.  I'm also curious how you decided to use 50 as the `LOOP_MAX`.

Also theres a .DS_STORE file change here that we should probably check in to .gitignore

yachtcaptain23 committed:
Improved padding calculation for image conversion.
First we calculate how much padding to add by creating a Tempfile with 1
character padding and checking the delta between the Tempfile's size and
the target file size.

Then we apply padding to the original file.

yachtcaptain23 committed:
Handle cases of file sizes being too big

yachtcaptain23 committed:
Document file size limits for the site banner

yachtcaptain23 committed:
Change resolution for publisher's logo to 240x240

yachtcaptain23 committed:
Ignore .DS_STORE file

yachtcaptain23 committed:
Support upscaled image size
